### ISSUE TITLE
Gradient Accumulation feature for Distribution Strategy

### ISSUE THREAD

**Issue by fanshiqing** [tensorflow/tensorflow#32176]
**System information**
- TensorFlow version (you are using): 1.12
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**
Gradient Accumulation(GA) is a workaround to enable big batches on limited memory GPUs which has been supported in Caffe and PyTorch. Instead of back-propagating for every batch feed-forward; gradients across multiple batches are accumulated. After multiple feed forwards, the accumulated gradient is back-propagated through the network layer.  It boosts performance with a couple of percentages on our several workload (e.g., XLNet, Transformer) with Distribution Strategy. Unfortunately, I am not aware of any official documentation to use such feature in Tensorflow. `tf.contrib.opt.AGNOptimizer` has a similar GA implementation but is not common for general distribution job on DistributionStrategy. Besides it will introduces OOM for large embeddings(#31637).
**Will this change the current api? How?**
Yes. Just need to add one new parameter `iter_size` to current strategy API, which  for example:
```
 # For MirroredStrategy, gradient accumulation is supported using iter_size parameter.
 distribution = tf.distribute.MirroredStrategy(num_gpus=2, iter_size=4)

 # For MultiWorkerMirroredStrategy:
 tf.distribute.experimental.MultiWorkerMirroredStrategy(..., iter_size=4)
```

**Who will benefit with this feature?**
Anyone who use distribution strategy and want to enable big batches on limited memory GPUs 
**Any Other info.**
None.

We already have a concise implementation of GA on DistributionStrategy and are willing to contribute. Thanks in advance for any feedback.

**Comment by yangjunpro** [comment]
Thank Shiqing for proposing this feature.
GA is not a fresh idea, the original idea may originated from 2012 in Microsoft's speech training acceleration paper, and in pyTorch and Caffe there is already alike implementation. Such as in 
https://gist.github.com/thomwolf/ac7a7da6b1888c2eeac8ac8b9b05d3d3
and 
https://github.com/BVLC/caffe/blob/128797ebe100dd87f375d2d30b0f649914d0fcc3/src/caffe/solver.cpp#L230  
https://github.com/BVLC/caffe/wiki/Solver-Prototxt#iter_size
This feature is quite useful since it brings two benefits:
1). Make it capable to train large models.  For example, if the original training could successfully train with batch size 1, then GA can boost the batch size to a much larger one. For lots of DL
model training, it is found that a decent batch size is better for model convergence than smaller batch size. So GA is useful for large model training.
2).Since large batch training techniques has been adopted for lots of scenarios, such as CV, NLP, speech and search&recommendation,  GA is an useful solution to increase computation-to-communication ration which in turn improves distributed scalability.  With GA, we can increase the iter_size to a extent with which the training statistical efficiency is acceptable(with tricks such as linear-learning rate rule or LARS, etc), thus greatly improve distributed training throughput.

We have already implemented GA on top of our internal version of TensorFlow and integrated it into CollectiveAllreduce and Mirrored strateg. Several users benefit from it with just one line of small code change. We think this feature may be useful for people outside of Alibaba so we raise this issue. If TF community like this idea, we can submit the PR very soon.

Thanks

**Comment by yangjunpro** [comment]
@yuefengz

**Comment by guillaumekln** [comment]
This is indeed a must have feature for recent sequence models. Could you clarify why you implemented this inside distribution strategies instead of as an optimizer wrapper?

**Comment by fanshiqing** [comment]
@guillaumekln Actually our GA implementation is an independent functional module which is called conditionally by base `optimizer` class. We add interface especially for distribution strategies as many of our models run top of it for its convenience and nice scalability.

**Comment by guillaumekln** [comment]
Interesting. Looking forward to see the code!

**Comment by ppwwyyxx** [comment]
A simple wrapper can be made to let any tensorflow 1.0 optimizer support Gradient Accumulation: https://github.com/tensorpack/tensorpack/blob/6a0bba688f2bac1c8baf59eb0c61eb90234a5424/tensorpack/tfutils/optimizer.py#L141-L224

So how is this related to distribution strategies? What types of integration with distribution strategies are you looking for that provide more benefits than simply wrapping the optimizer?

**Comment by yuefengz** [comment]
@fanshiqing Thank for your suggestion. I think this is an important and interesting feature, too. However, it would be difficult for us to change the distribution strategy's constructors. Other options may include changing the optimizer, or doing something in your [custom train loop](https://www.tensorflow.org/beta/tutorials/distribute/training_loops).

**Comment by fanshiqing** [comment]
@yuefengz Sorry for the delayed reply as I was on vocation. Add a new attribute to strategy's constructor is a direct approach. As what we want is just to pass a control-parameter like `iter_size` in some way, if the distribution strategy's constructors are difficult to change **explicitly**, what about add `iter_size` to `tf.estimator.RunConfig`, so that the original strategies' constructor will keep unchanged.

**Comment by ppwwyyxx** [comment]
If you just want to pass a parameter to `Optimizer`, then you should only change `Optimizer`. Because:
* Distribution strategy and Estimator do not even assume the existence of optimizer. They have their own scope of abstraction that are independent of what you do inside (which typically is an optimizer-based algorithm, but maybe not). Having an argument that's tied to specific types of common tasks is an abstraction leak (though I'm not saying that estimator does not have similar issues now).

* If you want to add an option to dist-strategy or `RunConfig` just to let them pass something to the optimizer, why not pass them to the optimizer directly? What benefits do you get by taking a longer path to achieve the same thing? In fact I only saw a disadvantage, which is:

* By adding such options to dist-strategy or `RunConfig` instead of to the optimizers directly, users who do now use dist-strategy or `RunConfig` (but still use optimizers) will not have access to this feature.

**Comment by fanshiqing** [comment]
@ppwwyyxx Thx for your nice advice. Actually at our scenario we only want to add GA for distribution strategy from the very begging. While it could be another story if we want to add this to community. It seems that pass control parameter to `Optimizer` directly is more reasonable. 
I will open a PR soon for more specific discussion.

**Comment by douglas125** [comment]
Is there a practical way to do this in tf.keras 2.x?

From a user perspective, having a parameter (e.g. in `.fit`) like `gradient_accumulate_batch_frequency=1` that updates gradients every N batches would be perfect. (1 as default to update every batch, as it is at the moment)

**Comment by Colanim** [comment]
+1, how can we achieve gradient accumulation in `tf.keras` (with `fit()` method) ?

**Comment by ethanyanjiali** [comment]
@ymodak @yuefengz Is there a timeline to add this gradient accumulation to tf.keras optimizers? Or will there be a guideline on how to do it in the custom training loop? My current approach is to create an accumulated gradients variable and have train step in each replica to add to this variable, but I also need a separate strategy.run to apply the gradients because that variable is under replica context.

**Comment by bhack** [comment]
@tomerk What do you want to do with the TF core at https://github.com/tensorflow/tensorflow/issues/32176

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this, as you are using an older version of tensorflow which is officially considered end of life . We recommend that you upgrade to the latest 2.x version and let us know if the issue still persists in newer versions. Please open a new issue for any help you need against 2.x, and we will get you the right help. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, please provide us with more information.

**Comment by bhack** [comment]
Check https://github.com/tensorflow/addons/pull/2525
